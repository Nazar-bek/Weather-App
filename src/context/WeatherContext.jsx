import React, { createContext, useContext, useEffect, useReducer } from "react";
import { initialState, weatherReducer } from "../reducers/weatherReducer";
import { useWeatherData } from "../hooks/useWeatherData";

export const WeatherContext = createContext();

const WeatherProvider = ({ children }) => {
  const [state, dispatch] = useReducer(weatherReducer, {
    ...initialState,
    unit: localStorage.getItem("unit") || initialState.unit
  });

    useEffect(() => {
    localStorage.setItem("unit", state.unit);
  }, [state.unit]);
  useWeatherData(state, dispatch)
  return (
    <WeatherContext.Provider value={{state, dispatch}}>
        {children}
    </WeatherContext.Provider>
  )
};

export const useWeather = () => useContext(WeatherContext)
export default WeatherProvider;