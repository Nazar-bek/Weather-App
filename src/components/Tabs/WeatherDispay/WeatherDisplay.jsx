import React from "react";
import { useWeather } from "../../../context/WeatherContext";

const WeatherDisplay = () => {
  const { state } = useWeather();
  const { weather, isLoading, error } = state;
  const unitSymbol = state.unit === "metric" ? "°C" : "°F";

  if (isLoading)
    return (
      <div className="flex items-center justify-center h-52">
        <div className="w-10 h-10 border-4 border-dashed rounded-full animate-spin border-blue-500"></div>
        <span className="ml-3 text-blue-500 font-medium text-lg">
          Loading...
        </span>
      </div>
    );

  if (error)
    return (
      <div className="flex items-center justify-center">
        <p className="text-red-500">{error}</p>
      </div>
    );

  if (!weather) return null;

  return (
    <div className="max-w-md mx-auto p-6 rounded-2xl shadow-lg bg-white dark:bg-[#1c1f23] text-gray-800 dark:text-gray-100 space-y-4">
      <h2 className="text-3xl font-bold">{weather.name}</h2>
      <p className="text-lg capitalize text-gray-600 dark:text-gray-300">
        {weather.weather[0].description}
      </p>

      <div className="flex justify-center">
        <img
          src={`https://openweathermap.org/img/wn/${weather.weather[0].icon}@2x.png`}
          alt={weather.name}
          className="w-24 h-24"
        />
      </div>

      <div className="grid grid-cols-1 sm:grid-cols-3 gap-4 text-sm text-center">
        <div className="bg-gray-100 dark:bg-[#2a2e33] rounded-xl p-4">
          <p className="font-medium text-lg">
            {Math.floor(weather.main.temp)}
            {unitSymbol}
          </p>
          <p className="text-gray-500 dark:text-gray-400">Temperature</p>
        </div>
        <div className="bg-gray-100 dark:bg-[#2a2e33] rounded-xl p-4">
          <p className="font-medium text-lg">{weather.main.humidity}%</p>
          <p className="text-gray-500 dark:text-gray-400">Humidity</p>
        </div>
        <div className="bg-gray-100 dark:bg-[#2a2e33] rounded-xl p-4">
          <p className="font-medium text-lg">{weather.wind.speed} m/s</p>
          <p className="text-gray-500 dark:text-gray-400">Wind</p>
        </div>
      </div>
    </div>
  );
};

export default WeatherDisplay;
