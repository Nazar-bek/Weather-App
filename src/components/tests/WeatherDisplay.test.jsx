import { describe, it, expect, vi } from "vitest";
import { render } from "@testing-library/react";
import React from "react";
import WeatherDisplay from "../Tabs/WeatherDispay/WeatherDisplay";
import { WeatherContext } from "../../context/WeatherContext";

describe("WeatherDisplay snapshot test", () => {
  it("matches snapshot", () => {
    const mockState = {
      weather: {
        name: "Tashkent",
        main: {
          temp: 28,
          feels_like: 30,
          temp_min: 25,
          temp_max: 32,
        },
        weather: [
          {
            description: "clear sky",
            icon: "01d",
          },
        ],
        wind: {
          speed: 4.5, 
        },
      },
      isLoading: false,
      error: null,
      unit: "metric",
    };

    const mockDispatch = vi.fn();

    const { asFragment } = render(
      <WeatherContext.Provider value={{ state: mockState, dispatch: mockDispatch }}>
        <WeatherDisplay />
      </WeatherContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
