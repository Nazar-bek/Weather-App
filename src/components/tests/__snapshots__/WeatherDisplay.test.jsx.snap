// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeatherDisplay snapshot test > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-md mx-auto p-6 rounded-2xl shadow-lg bg-white dark:bg-[#1c1f23] text-gray-800 dark:text-gray-100 space-y-4"
  >
    <h2
      class="text-3xl font-bold"
    >
      Tashkent
    </h2>
    <p
      class="text-lg capitalize text-gray-600 dark:text-gray-300"
    >
      clear sky
    </p>
    <div
      class="flex justify-center"
    >
      <img
        alt="Tashkent"
        class="w-24 h-24"
        src="https://openweathermap.org/img/wn/01d@2x.png"
      />
    </div>
    <div
      class="grid grid-cols-1 sm:grid-cols-3 gap-4 text-sm text-center"
    >
      <div
        class="bg-gray-100 dark:bg-[#2a2e33] rounded-xl p-4"
      >
        <p
          class="font-medium text-lg"
        >
          28Â°C
        </p>
        <p
          class="text-gray-500 dark:text-gray-400"
        >
          Temperature
        </p>
      </div>
      <div
        class="bg-gray-100 dark:bg-[#2a2e33] rounded-xl p-4"
      >
        <p
          class="font-medium text-lg"
        >
          %
        </p>
        <p
          class="text-gray-500 dark:text-gray-400"
        >
          Humidity
        </p>
      </div>
      <div
        class="bg-gray-100 dark:bg-[#2a2e33] rounded-xl p-4"
      >
        <p
          class="font-medium text-lg"
        >
          4.5 m/s
        </p>
        <p
          class="text-gray-500 dark:text-gray-400"
        >
          Wind
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
